Fixes issue #40: Adds Generic Documents for YAML and Config (#64)
aquasecurity/bench-common: docs/README.md

      audit: "ps -ef | grep kube-apiserver | grep -v grep"
      tests:
      bin_op: or
      test_items:
      - flag: "--allow-privileged"
        set: true
      - flag: "--some-other-flag"
        set: false
      remediation: "Edit the /etc/kubernetes/config file on the master node and
        set the KUBE_ALLOW_PRIV parameter to '--allow-privileged=false'"
      scored: true
- id: 1.2
  text: Scheduler
  checks:
    - id: 1.2.1
      text: "Ensure that the --profiling argument is set to false (Scored)"
      audit: "ps -ef | grep kube-scheduler | grep -v grep"
      tests:
        bin_op: or
        test_items:
          - flag: "--profiling"
            set: true
          - flag: "--some-other-flag"
            set: false
      remediation: "Edit the /etc/kubernetes/config file on the master node and
        set the KUBE_ALLOW_PRIV parameter to '--allow-privileged=false'"
      scored: true
```

`controls` is composed of a hierarchy of groups, sub-groups and checks. Each of
the `controls` components have an id and a text description which are displayed 
in the `*-bench` project output.

`type` specifies the type a `controls` is for.

## Groups

> `groups` is list of subgroups which test the various Kubernetes components

Reviewer: Kubernetes reference in generic text

    audit: "ps -ef | grep kube-apiserver | grep -v grep"
    tests:
    # ...
``` 

The `*-bench` project supports running a subgroup by specifying the subgroup `id` on the
command line, with the flag `--group` or `-g`.

## Check

In a `*-bench` project , a `check` object embodies a recommendation from the CIS benchmark.  This an example
`check` object:

```yml
id: 1.1.1
text: "Ensure that the --anonymous-auth argument is set to false (Not Scored)"
audit: "ps -ef | grep kube-apiserver | grep -v grep"
tests:
  test_items:
  - flag: "--anonymous-auth"
    compare:
      op: eq
      value: false
    set: true
remediation: |
  Edit the API server pod specification file kube-apiserver
  on the master node and set the below parameter.
  --anonymous-auth=false
scored: false
```

A `check` object has an `id`, a `text`, an `audit`, a `tests`, `remediation`
and `scored` fields.

The `*-bench` project supports running individual checks by specifying the check's `id`
as a comma-delimited list on the command line with the `--check` flag.

The `audit` field specifies the command to run for a check. The output of this

> command is then evaluated for conformance with the CIS Kubernetes Benchmark

Reviewer: ```suggestion
command is then evaluated for conformance with the CIS Benchmark
```

> false

The `op` (operations) currently supported in The `*-bench` project are:
- `eq`: tests if the keyword is equal to the compared value.
- `noteq`: tests if the keyword is unequal to the compared value.
- `gt`: tests if the keyword is greater than the compared value.
- `gte`: tests if the keyword is greater than or equal to the compared value.
- `lt`: tests if the keyword is less than the compared value.
- `lte`: tests if the keyword is less than or equal to the compared value.
- `has`: tests if the keyword contains the compared value.
- `nothave`: tests if the keyword does not contain the compared value.
- `valid_elements`: tests if the keyword contains valid elements from the list of values provided.
  The values in the list provided uses a `,`  as a separator.
- `regex`: tests if the flag value matches the compared value regular expression.
   When defining regular expressions in YAML it is generally easier to wrap them in
   single quotes, for example `'^[abc]$'`, to avoid issues with string escaping.

## Configuration and Variables

Ccomponent configuration and binary file locations and names 

> vary based on cluster deployment methods, Operating Systems, and Kubernetes distribution used.

Reviewer: Another Kubernetes reference

# Test and config files

A typical `*-bench` project runs checks specified in `controls` files that are a YAML 
representation of the CIS Benchmark checks. There is a 

> `controls` file per version and node type.

Reviewer: The second sentence is Kubernetes-specific (I don't think all benchmarks have node type)

> `controls` for the various versions can be found in directories

Reviewer: ```suggestion
`controls` for the various versions of the benchmark can be found in directories
```

with same name as the CIS versions under `cfg/`, for example `cfg/cis-1.4.0`.

> `controls` are also organized by distribution under the `cfg` directory, for

Reviewer: Also not sure that "organized by distribution" is generic, I would drop that sentence

example `cfg/rh-0.7`.

## Controls

`controls` is a YAML document that contains checks that must be run against a 
specific Kubernetes node type, master or node and version.

> 

Reviewer: Again Kubernetes specific. Maybe better: "`controls` is a YAML document containing the test definitions for a benchmark"

`controls` is the fundamental input to a `*-bench` project. The following is a Kubernetes CIS Benchmark example 

> of a basic `controls`:

Reviewer: ```suggestion
of `controls`:
```

      audit: "ps -ef | grep kube-apiserver | grep -v grep"
      tests:
      bin_op: or
      test_items:
      - flag: "--allow-privileged"
        set: true
      - flag: "--some-other-flag"
        set: false
      remediation: "Edit the /etc/kubernetes/config file on the master node and
        set the KUBE_ALLOW_PRIV parameter to '--allow-privileged=false'"
      scored: true
- id: 1.2
  text: Scheduler
  checks:
    - id: 1.2.1
      text: "Ensure that the --profiling argument is set to false (Scored)"
      audit: "ps -ef | grep kube-scheduler | grep -v grep"
      tests:
        bin_op: or
        test_items:
          - flag: "--profiling"
            set: true
          - flag: "--some-other-flag"
            set: false
      remediation: "Edit the /etc/kubernetes/config file on the master node and
        set the KUBE_ALLOW_PRIV parameter to '--allow-privileged=false'"
      scored: true
```

`controls` is composed of a hierarchy of groups, sub-groups and checks. Each of
the `controls` components have an id and a text description which are displayed 
in the `*-bench` project output.

`type` specifies what kubernetes node type a `controls` is for. Possible values

> for `type` are `master` and `node`.

Reviewer: This is Kubernetes-specific.

## Groups

`groups` is list of subgroups which test the various Kubernetes components

> that run on the node type specified in the `controls`. 

Reviewer: Kubernetes specific. I think we could make this generic by saying "`groups` represents a section in the benchmark. Depending on the benchmark a section (and thereby its group in the YAML definition` may represent a component, a group of components, or a set of functionality."

Reviewer: This still refers to Kubernetes

For example one subgroup checks parameters passed to the apiserver binary, while 

> another subgroup checks parameters passed to the controller-manager binary.

Reviewer: I would just delete this, it's confusing in a generic document

in the previous paragraphs. The most important part of the subgroup is the
`checks` field which is the collection of actual `check`s that form the subgroup.

This is an example of a subgroup and checks in the subgroup.

```yml
id: 1.1
text: API Server
checks:
  - id: 1.1.1
    text: "Ensure that the --allow-privileged argument is set (Scored)"
    audit: "ps -ef | grep kube-apiserver | grep -v grep"
    tests:
    # ...
  - id: 1.1.2
    text: "Ensure that the --anonymous-auth argument is set to false (Not Scored)"
    audit: "ps -ef | grep kube-apiserver | grep -v grep"
    tests:
    # ...
``` 

The `*-bench` project supports running a subgroup by specifying the subgroup `id` on the
command line, with the flag `--group` or `-g`.

## Check

The CIS Benchmark recommends configurations to harden various 
components, including Docker, Kubernetes, Linux, etc... 
These recommendations are usually configuration options, and can be 
specified by flags to binaries programs, or in configuration files.

The Benchmark also provides commands to audit an installation, identify
places where the security can be improved, and steps to remediate these

> identified problems.

Reviewer: This feels like an introductory section that should be at the top of the file?

> In a `*-bench` project , `check` objects embody these recommendations.  This an example

Reviewer: ```suggestion
In a `*-bench` project , a `check` object embodies a recommendation from the CIS benchmark.  This an example
```

  `check`.
  
  ```yml
  id: 2.1.1
    # ...
    remediation: |
      Edit the kubelet service file $kubeletsvc
      on each worker node and set the below parameter in KUBELET_SYSTEM_PODS_ARGS variable.
      --allow-privileged=false
      Based on your system, restart the kubelet service. For example:
      systemctl daemon-reload
      systemctl restart kubelet.service
    # ...
  ```
  
  - `kubeconfig`: A list of candidate kubeconfig files for a component. The `*-bench` project
    checks this list and selects the first file that is found on the node. If none
    of the files exists, The `*-bench` project defaults kubeconfig to the value of 
    `defaultkubeconfig`.
    
    The selected kubeconfig for a component can be referenced in `controls` with
    a variable in the form `$<component>kubeconfig`. In the example below, the
    selected kubelet kubeconfig is referenced with `$kubeletkubeconfig` in the
    `audit` command.
    
    ```yml
    id: 2.2.1
      text: "Ensure that the kubelet.conf file permissions are set to 644 or
      more restrictive (Scored)"
      audit: "/bin/sh -c 'if test -e $kubeletkubeconfig; then stat -c %a $kubeletkubeconfig; fi'"
      # ...

>     ```

Reviewer: I think this whole section only applies to kube-bench, not bench-common
